### PR TITLE
[ADD] sale_order_team_from_product: add module

### PR DESCRIPTION
**Pull request by tendil**
- Add settings flags to control taking Sales Team from products and skipping when already set.
- Compute team from product.template.sales_team_id on sale.order create/write for draft/sent orders.
- Avoid recursion via `sale_order_team_from_product_skip` context key.
- Add demo team, reuse existing demo products with sales_team_id, and demo sale order.
- Add unit tests for configuration, team computation, create/write behavior and conflict cases.

Task: 5085

## Release Notes

* **New Features**
  * Added automatic sales team assignment to quotations based on product team definitions.
  * Added configuration settings to enable the feature and control whether existing team assignments are preserved.

* **Documentation**
  * Added comprehensive module documentation including configuration, usage guidance, and roadmap.
  * Added demo data showcasing the feature functionality.

* **Tests**
  * Added unit tests covering configuration flags, team assignment logic, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: sale_order_team_from_product/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        """Update sale orders and re-apply Sales Team from products when relevant."""
+        res = super().write(vals)
+        if self.env.context.get("sale_order_team_from_product_skip"):
+            return res
+        if not {"order_line", "team_id"} & set(vals.keys()):
+            return res
+        enabled, skip_if_set = self._get_sales_team_from_product_config()
+        if enabled:
+            self._apply_sales_team_from_products(skip_if_set)
+        return res
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py]: That's how it's supposed to be, because you have to check for changes so as not to cause unnecessary problems

File: sale_order_team_from_product/readme/CONTEXT.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Sometimes you need to set a product specific sales team in quotation.
+Eg the product you are selling requires some specific knowledge and thus a dedicated team must be in charge of the sales process.
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/readme/CONTEXT.md:2]: That's how it was written in the task so no changes are needed

File: sale_order_team_from_product/readme/CONTRIBUTORS.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+* [Cetmix](https://cetmix.com/):
+  * Ivan Sokolov
+  * Dmitry Meita
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/readme/CONTRIBUTORS.md:3]: Everything is fine here

File: sale_order_team_from_product/readme/CREDITS.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+The development of this module has been financially supported by:
+  - Geschäftsstelle Sozialinfo
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/readme/CREDITS.md:2]: Everything is fine here

File: test-requirements.txt [hunk 5]
```diff
@@ -1 +1,2 @@
 odoo-test-helper
+odoo-addon-product_sale_team @ git+https://github.com/OCA/product-attribute@refs/pull/2139/head#subdirectory=setup/product_sale_team
```
[the remark this answers is not in the document]
Comment (tendil) [reply test-requirements.txt]: I changed the path and specified without setup/, and everything is fine now

File: sale_order_team_from_product/tests/test_sale_order_team_from_product.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        self.assertFalse(team)
+
+    def test_create_no_feature_enabled_does_not_change_team(self):
+        """When feature disabled, create must not change team."""
+        self._set_config(enabled=False, skip_if_set=False)
+        order = self._create_order([self.product_team_a_1, self.product_team_a_2])
+        default_team = self.env.ref("sales_team.team_sales_department")
+        # Team stays whatever Odoo default sets (False, unless other logic)
+        self.assertEqual(order.team_id, default_team)
+
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/tests/test_sale_order_team_from_product.py:172]: The tests are run on a clean database containing demo data, and the order will always contain the command by default

File: sale_order_team_from_product/tests/test_sale_order_team_from_product.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        order = self._create_order([self.product_team_a_1, self.product_team_a_2])
+        team = order._get_sales_team_from_products()
+        self.assertEqual(team, self.team_a)
+
+    def test_get_sales_team_from_products_different_teams(self):
+        """If products have different teams, method must return False."""
+        order = self._create_order([self.product_team_a_1, self.product_team_b_1])
+        team = order._get_sales_team_from_products()
+        self.assertFalse(team)
+
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/tests/test_sale_order_team_from_product.py:164]: no need

File: test-requirements.txt [hunk 8]
```diff
@@ -1 +1,2 @@
 odoo-test-helper
+odoo-addon-product_sale_team @ git+https://github.com/OCA/product-attribute@refs/pull/2139/head#subdirectory=product_sale_team
```
[the remark this answers is not in the document]
Comment (tendil) [reply test-requirements.txt]: Everything is fine here the tests are running smoothly

File: sale_order_team_from_product/demo/sale_order_demo.xml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        <field name="partner_invoice_id" ref="base.res_partner_2" />
+        <field name="partner_shipping_id" ref="base.res_partner_2" />
+        <field name="user_id" ref="base.user_demo" />
+        <field
+            name="team_id"
+            ref="sale_order_team_from_product.crm_team_from_product_demo"
+        />
+        <field name="source_id" ref="sale.utm_source_sale_order_0" />
+        <field name="tag_ids" eval="[(4, ref('sales_team.categ_oppor4'))]" />
+    </record>
```
Comment (tendil) [inline sale_order_team_from_product/demo/sale_order_demo.xml:58]: no need

File: sale_order_team_from_product/demo/sale_order_demo.xml [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        <field name="name">Team from Product (Demo)</field>
+        <field name="company_id" ref="base.main_company" />
+    </record>
+
+    <function model="ir.model.data" name="_update_xmlids">
+        <value
+            model="base"
+            eval="[
+            {
+                'xml_id': 'sale_order_team_from_product.product_product_1_product_template',
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/demo/sale_order_demo.xml]: External demo data from the modules we don't control may change in the future. This makes module maintenance and migration more complicated.
Why not to create a couple of own products and use them?

File: sale_order_team_from_product/models/res_config_settings.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    skip_sales_team_if_set = fields.Boolean(
+        string="Skip if team is already set",
+        help=(
+            "If enabled, the Sales Team from products will not override an "
+            "already set Sales Team on the quotation."
+        ),
+    )
+
+    @api.model
+    def get_values(self):
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/res_config_settings.py]: This is handled by the ORM automatically with the corresponding [key](https://github.com/odoo/odoo/blob/4b897c262f94d7adc03f21065c4210fa77f7ae2f/addons/account_payment/wizards/res_config_settings.py#L9) in the field definition.

File: sale_order_team_from_product/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+from odoo.tools import str2bool
+
+from .res_config_settings import _PARAM_SET_FROM_PRODUCTS, _PARAM_SKIP_IF_SET
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.model_create_multi
+    def create(self, vals_list):
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/sale_order.py]: [team_id](https://github.com/odoo/odoo/blob/b6b79cb96b7d30e6d3213d8b65d2be1a5b17b32d/addons/sale/models/sale_order.py#L223) is a computed field. Please check the source code you are modifying **before** writing your code.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: sale_order_team_from_product/models/res_config_settings.py (hunk 11 above)
Comment (tendil) [reply sale_order_team_from_product/models/res_config_settings.py]: Yes
I initially had this implementation, but there was an error in that when the nested setting was disabled it was not canceled when saving, so I decided to use an explicit method

I will double-check, thank you

File: sale_order_team_from_product/demo/sale_order_demo.xml [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+    <record id="sale_order_line_demo_team_product_2" model="sale.order.line">
+        <field name="order_id" ref="sale_order_demo_team_from_product" />
+        <field
+            name="product_id"
+            ref="sale_order_team_from_product.product_team_from_product_2"
+        />
+        <field name="product_uom_qty">1.0</field>
+    </record>
+</odoo>
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/demo/sale_order_demo.xml:77]: no need

File: sale_order_team_from_product/models/sale_order.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        for order in self:
+            # We do not touch confirmed orders
+            if order.state not in ("draft", "sent"):
+                continue
+            if skip_if_set and order.team_id:
+                continue
+            team = order._get_sales_team_from_products()
+            if not team:
+                continue
+            order.team_id = team
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py]: Not required
documentation is already in order

File: sale_order_team_from_product/demo/sale_order_demo.xml [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        <field name="partner_invoice_id" ref="base.res_partner_2" />
+        <field name="partner_shipping_id" ref="base.res_partner_2" />
+        <field name="user_id" ref="base.user_demo" />
+        <field
+            name="team_id"
+            ref="sale_order_team_from_product.crm_team_from_product_demo"
+        />
+        <field name="source_id" ref="sale.utm_source_sale_order_0" />
+        <field name="client_order_ref">Demo - Sales Team from Product</field>
+    </record>
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/demo/sale_order_demo.xml:58]: no need

File: sale_order_team_from_product/demo/sale_order_demo.xml [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        <field name="list_price">20.0</field>
+        <field name="uom_id" ref="uom.product_uom_unit" />
+        <field name="uom_po_id" ref="uom.product_uom_unit" />
+        <field
+            name="sales_team_id"
+            ref="sale_order_team_from_product.crm_team_from_product_demo"
+        />
+    </record>
+
+    <function model="ir.model.data" name="_update_xmlids">
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/demo/sale_order_demo.xml:31]: Why do we need this function?

File: sale_order_team_from_product/models/sale_order.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+          order lines.
+        - Ignore empty values.
+        - If exactly one unique team is found, return that team recordset.
+        - If there are no teams or more than one distinct team, return False.
+
+        :return: A crm.team recordset with exactly one record, or False.
+        """
+        self.ensure_one()
+        teams = self.order_line.mapped("product_id.product_tmpl_id.sales_team_id")
+        teams = teams.filtered("id")
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/sale_order.py]: Please explain this operation.

File: sale_order_team_from_product/models/sale_order.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        - Collect sales_team_id from related product templates on all
+          order lines.
+        - Ignore empty values.
+        - If exactly one unique team is found, return that team recordset.
+        - If there are no teams or more than one distinct team, return False.
+
+        :return: A crm.team recordset with exactly one record, or False.
+        """
+        self.ensure_one()
+        teams = self.order_line.mapped("product_id.product_tmpl_id.sales_team_id")
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/sale_order.py]: You can use dot-notation for x2m fields:
```python
self.order_line.product_id.product_tmpl_id.sales_team_id
```

File: sale_order_team_from_product/models/sale_order.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+            if skip_if_set and order.team_id:
+                continue
+            team = order._get_sales_team_from_products()
+            if not team:
+                continue
+            order.team_id = team
+
+    @api.depends(
+        "partner_id",
+        "user_id",
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/sale_order.py]: These keys are already [present](https://github.com/odoo/odoo/blob/b6b79cb96b7d30e6d3213d8b65d2be1a5b17b32d/addons/sale/models/sale_order.py#L476) in the depends of the original method, no need to add them again.

File: sale_order_team_from_product/models/sale_order.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        if the feature is enabled, optionally replaces the result with the
+        team coming from the products on the order.
+        """
+        # First, let the core logic compute the default team.
+        super()._compute_team_id()
+
+        # Then, optionally override it using product configuration.
+        enabled, skip_if_set = self._get_sales_team_from_product_config()
+        if not enabled:
+            return
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/sale_order.py]: ??

File: sale_order_team_from_product/static/description/icon.png
Comment (ivs-cetmix) [inline sale_order_team_from_product/static/description/icon.png:1]: This is the old icon, pls remove it. Icon will be added automatically when the PR is merged.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: sale_order_team_from_product/demo/sale_order_demo.xml (hunk 16 above)
Comment (tendil) [reply sale_order_team_from_product/demo/sale_order_demo.xml:31]: we need this function because the products we use in the demo are defined on `product.template`, and odoo automatically creates the corresponding `product.product` variants without any xml ids

this function lets us assign stable xml ids to those automatically created variants, so that we can reference them in the sale order lines instead of using brittle `eval` expressions or creating extra variants manually

File: sale_order_team_from_product/models/sale_order.py (hunk 17 above)
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py]: - `teams = teams.filtered("id")` removes any empty values (lines whose product template has no `sales_team_id` at all).

After that we simply check:
- if there is no team > return `False`
- if there is exactly one unique team > return that team
- if there are multiple different teams > also return `False` to avoid choosing arbitrarily

File: sale_order_team_from_product/models/sale_order.py (hunk 18 above)
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py]: thanks! 
i’ve replaced the mapped() usage with dot-notation

File: sale_order_team_from_product/models/sale_order.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        team coming from the products on the order.
+        """
+        # First, let the core logic compute the default team.
+        res = super()._compute_team_id()
+
+        # Then, optionally override it using product configuration.
+        enabled, skip_if_set = self._get_sales_team_from_product_config()
+        if enabled:
+            self._apply_sales_team_from_products(skip_if_set)
+        return res
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py]: Without this, we will not pass the pre-commit check, and there will be an error:
- `sale_order_team_from_product/models/sale_order.py:81: [W8110(missing-return), SaleOrder._compute_team_id] Missing `return` (`super` is used) in method _compute_team_id.`

[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py]: For clarity, I removed the variable declaration and simply suppressed the error - # pylint: disable=W8110

File: sale_order_team_from_product/demo/sale_order_demo.xml (hunk 16 above)
Comment (ivs-cetmix) [reply sale_order_team_from_product/demo/sale_order_demo.xml:31]: Noted, thank you!

File: sale_order_team_from_product/demo/sale_order_demo.xml [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+            eval="[
+            {
+                'xml_id': 'sale_order_team_from_product.product_team_from_product_1',
+                'record': obj().env.ref('sale_order_team_from_product.product_team_from_product_tmpl_1').product_variant_id,
+                'noupdate': True,
+            },
+            {
+                'xml_id': 'sale_order_team_from_product.product_team_from_product_2',
+                'record': obj().env.ref('sale_order_team_from_product.product_team_from_product_tmpl_2').product_variant_id,
+                'noupdate': True,
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/demo/sale_order_demo.xml]: You put "noupdate: True" for the product variant, however product template is created with "noupdate: False"

File: sale_order_team_from_product/models/sale_order.py (hunk 17 above)
Comment (ivs-cetmix) [reply sale_order_team_from_product/models/sale_order.py]: You can use the ids instead of records, something like this:

```python
team_ids = set(self.order_line.product_id.product_tmpl_id.sales_team_id.ids)
if len(team_ids) == 1 and team_ids[0]:
...you have a single valid team...
```

And you can return a team_id instead of a team records.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: sale_order_team_from_product/models/sale_order.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+        team coming from the products on the order.
+        """
+        # pylint: disable=W8110
+        # First, let the core logic compute the default team.
+        super()._compute_team_id()
+
+        # Then, optionally override it using product configuration.
+        enabled, skip_if_set = self._get_sales_team_from_product_config()
+        if enabled:
+            self._apply_sales_team_from_products(skip_if_set)
```
[the remark this answers is not in the document]
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py:95]: no need
everything is fine here

File: sale_order_team_from_product/models/sale_order.py [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        Base implementation computes the default Sales Team from the
+        salesperson, partner and company.
+
+        This override keeps the original behavior and,
+        if the feature is enabled, optionally replaces the result with the
+        team coming from the products on the order.
+        """
+        # pylint: disable=W8110
+        # First, let the core logic compute the default team.
+        super()._compute_team_id()
```
Comment (ivs-cetmix) [inline sale_order_team_from_product/models/sale_order.py:90]: Do we need those `super` values in the `_get_sales_team_from_product_config()`? 
If no, then no need to fetch them if they will be replaced later anyway.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: sale_order_team_from_product/models/sale_order.py (hunk 24 above)
Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py:90]: do you mean _compute_team_id or _get_sales_team_from_product_config ?

Comment (tendil) [reply sale_order_team_from_product/models/sale_order.py:90]: in our case we don't always obscure the meanings and this is an important point

`super()._compute_team_id()` is needed because:
- When the feature is turned off, `_apply_sales_team_from_products` is not called at all >
we rely entirely on the standard calculation of team_id.

Comment (ivs-cetmix) [reply sale_order_team_from_product/models/sale_order.py:90]: Ok, I see.

**Review by ivs-cetmix — APPROVED** [review]

**Comment by ivs-cetmix** [comment]
@tendil pls check the comments and apply fixes if needed
